ocaml/libs/tracing/tracing.ml

```diff
@@ -0,0 +1,674 @@
+          Xapi_stdext_unix.Unixext.write_string_to_file file span_json ;
+          Ok ""
+        with e -> Error e
+    end
+
+    module Http = struct
+      module Request = Cohttp.Request.Make (Cohttp_posix_io.Buffered_IO)
+      module Response = Cohttp.Response.Make (Cohttp_posix_io.Buffered_IO)
+
+      let export ~span_json ~url : (string, exn) result =
+        try
+          let body = span_json in
+          let uri = Uri.of_string url in
+          let headers =
+            Cohttp.Header.of_list
+              [
+                ("accepts", "application/json")
+              ; ("content-type", "application/json")
+              ; ("content-length", string_of_int (String.length body))
+              ]
+          in
+          Open_uri.with_open_uri uri (fun fd ->
+              let request =
+                Cohttp.Request.make ~meth:`POST ~version:`HTTP_1_1 ~headers uri
+              in
+              let ic = Unix.in_channel_of_descr fd in
+              let oc = Unix.out_channel_of_descr fd in
+              Request.write
+                (fun writer -> Request.write_body writer body)
+                request oc ;
+              Unix.shutdown fd Unix.SHUTDOWN_SEND ;
+              match try Response.read ic with _ -> `Eof with
+              | `Eof ->
+                  Ok ""
+              | `Invalid x ->
+                  Error (Failure ("invalid read: " ^ x))
+              | `Ok response ->
+                  let body = Buffer.create 128 in
+                  let reader = Response.make_body_reader response ic in
+                  let rec loop () =
+                    match Response.read_body_chunk reader with
+                    | Cohttp.Transfer.Chunk x ->
+                        Buffer.add_string body x ; loop ()
+                    | Cohttp.Transfer.Final_chunk x ->
+                        Buffer.add_string body x
+                    | Cohttp.Transfer.Done ->
+                        ()
+                  in
+                  loop () ;
+                  Ok (Buffer.contents body)
+          )
+        with e -> Error e
+    end
+
+    let export_to_endpoint endpoint =
+      try
+        debug "Tracing: About to export" ;
+        let span_list = Spans.since () in
+        Hashtbl.iter
+          (fun trace_id span_list ->
+            let zipkin_spans = Content.Json.Zipkinv2.content_of span_list in
+            match
+              match endpoint with
+              | Url uri ->
+                  let url = Uri.to_string uri in
+                  Http.export ~span_json:zipkin_spans ~url
+              | Bugtool ->
+                  File.export ~trace_id ~span_json:zipkin_spans
+                    ~path:!File.trace_log_dir
+            with
+            | Ok _ ->
+                ()
+            | Error e ->
+                raise e
+          )
+          span_list
+      with e -> debug "Tracing: ERROR %s" (Printexc.to_string e)
+
+    let _ =
```

Comment:
We need the export thread for multiple daemons as well, this was it is done without needing to start this up in all of them.

Reply:
For libraries especially it is better to have an init function rather than a `let _ =` "main" function, to give the users of the library more control. Having multiple "main" functions in a program just adds unpredictability and risks race conditions.